// @yarnpkg/pnpify/lib/sdks/vscode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VSCODE_SDKS = exports.generateSvelteLanguageServerWrapper = exports.generateStylelintWrapper = exports.generateTypescriptWrapper = exports.generatePrettierWrapper = exports.generateEslintWrapper = exports.addVSCodeWorkspaceConfiguration = exports.VSCodeConfiguration = exports.merge = void 0;
const tslib_1 = require("tslib");
const fslib_1 = require("@yarnpkg/fslib");
const mergeWith_1 = tslib_1.__importDefault(require("lodash/mergeWith"));
const sdkUtils = tslib_1.__importStar(require("../sdkUtils"));
exports.merge = (object, source) => mergeWith_1.default(object, source, (objValue, srcValue) => {
    if (Array.isArray(objValue))
        return [...new Set(objValue.concat(srcValue))];
    return undefined;
});
var VSCodeConfiguration;
(function (VSCodeConfiguration) {
    VSCodeConfiguration["settings"] = "settings.json";
    VSCodeConfiguration["extensions"] = "extensions.json";
})(VSCodeConfiguration = exports.VSCodeConfiguration || (exports.VSCodeConfiguration = {}));
exports.addVSCodeWorkspaceConfiguration = async (pnpApi, type, patch) => {
    const relativeFilePath = `.vscode/${type}`;
    await sdkUtils.addSettingWorkspaceConfiguration(pnpApi, relativeFilePath, patch);
};
exports.generateEslintWrapper = async (pnpApi, target, wrapper) => {
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.settings, {
        [`eslint.nodePath`]: fslib_1.npath.fromPortablePath(fslib_1.ppath.dirname(fslib_1.ppath.dirname(fslib_1.ppath.dirname(wrapper.getProjectPathTo(`lib/api.js`))))),
    });
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `dbaeumer.vscode-eslint`,
        ],
    });
};
exports.generatePrettierWrapper = async (pnpApi, target, wrapper) => {
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.settings, {
        [`prettier.prettierPath`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(`index.js`)),
    });
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `esbenp.prettier-vscode`,
        ],
    });
};
exports.generateTypescriptWrapper = async (pnpApi, target, wrapper) => {
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.settings, {
        [`typescript.tsdk`]: fslib_1.npath.fromPortablePath(fslib_1.ppath.dirname(wrapper.getProjectPathTo(`lib/tsserver.js`))),
        [`typescript.enablePromptUseWorkspaceTsdk`]: true,
    });
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `arcanis.vscode-zipfs`,
        ],
    });
};
exports.generateStylelintWrapper = async (pnpApi, target, wrapper) => {
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.settings, {
        [`stylelint.stylelintPath`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(`lib/index.js`)),
    });
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `stylelint.vscode-stylelint`,
        ],
    });
};
exports.generateSvelteLanguageServerWrapper = async (pnpApi, target, wrapper) => {
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.settings, {
        [`svelte.language-server.ls-path`]: fslib_1.npath.fromPortablePath(wrapper.getProjectPathTo(`bin/server.js`)),
    });
    await exports.addVSCodeWorkspaceConfiguration(pnpApi, VSCodeConfiguration.extensions, {
        [`recommendations`]: [
            `svelte.svelte-vscode`,
        ],
    });
};
exports.VSCODE_SDKS = [
    [`eslint`, exports.generateEslintWrapper],
    [`prettier`, exports.generatePrettierWrapper],
    [`typescript-language-server`, null],
    [`typescript`, exports.generateTypescriptWrapper],
    [`stylelint`, exports.generateStylelintWrapper],
    [`svelte-language-server`, exports.generateSvelteLanguageServerWrapper],
];
